Add new ConfigDocument API methods

fpringvaldsen opened the pull request:
Add two new methods to the ConfigDocument API, hasValue() and
removeValue(), along with tests.

havocp commented on config/src/main/java/com/typesafe/config/ConfigDocument.java:
```diff
@@ -51,6 +51,24 @@
+     * @param path the path to remove from the document
+     * @return a copy of the ConfigDocument with the desired value removed from the document.
+     */
+    ConfigDocument removeValue(String path);
```
docs should say "all values at the desired path" probably (if there are multiple, we remove all presumably)

fpringvaldsen commented:
I need to rebase and make a few other small changes, but this should at least be ready for review now.

havocp commented on config/src/main/java/com/typesafe/config/ConfigDocument.java:
```diff
@@ -51,6 +51,24 @@
+    ConfigDocument removeValue(String path);
+
+    /**
+     * Returns a boolean indicating whether or not a ConfigDocument has a value at the desired path.
```
maybe add "Substitutions (the `<code>${foo}</code>` syntax) are always considered to be values for purposes of this method, even though they might disappear once the file is resolved." 
Or something like that.

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigNodeObject.java:
```diff
@@ -12,32 +12,83 @@
-    protected ConfigNodeObject changeValueOnPath(Path desiredPath, AbstractConfigNodeValue value) {
+    public boolean hasValue(Path desiredPath) {
+        for (AbstractConfigNode node : children) {
+            if (node instanceof ConfigNodeField) {
```
if you put `ConfigNodeField field = (ConfigNodeField) node` here and then use `field` below, then you could avoid some of the cast noise in the code below.

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigNodeObject.java:
```diff
@@ -12,32 +12,83 @@
+                if (key.equals(desiredPath) || key.startsWith(desiredPath)) {
+                    return true;
+                } else if (desiredPath.startsWith(key)) {
+                    if (((ConfigNodeField) node).value() instanceof ConfigNodeObject) {
```
similarly here you could write `ConfigNodeObject obj = (ConfigNodeObject) field.value()` to avoid mixing cast noise into the logic below.

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigNodeObject.java:
```diff
@@ -12,32 +12,83 @@
+                    childrenCopy.remove(i);
+                }
+                continue;
+            }else if (!(childrenCopy.get(i) instanceof ConfigNodeField)) {
```
missing a space after the `}`

havocp commented on config/src/test/scala/com/typesafe/config/impl/ConfigDocumentTest.scala:
```diff
@@ -193,6 +225,28 @@ class ConfigDocumentTest extends TestUtils {
+                assertTrue(e.isInstanceOf[ConfigException])
+                assertTrue(e.getMessage.contains("ConfigDocument had an array at the root level"))
+        }
+        assertTrue(exceptionThrown)
```
you can use `val e = intercept[ConfigException] { document.hasValue("a") }; assertTrue(e.getMessage.contains)` here, which will check the same things with less code. (intercept[E] throws if the exception is not intercepted or doesn't have the expected type)

havocp commented:
will need fixing up since I moved files around, too. (sorry about that)

fpringvaldsen commented:
@havocp No problem about the package stuff! I'm going to rebase now. I just added a new commit adding a couple new tests and changing the way duplicate removal works in the ConfigDocument, and after lunch I'll get to work on addressing your comments.

fpringvaldsen commented:
Thanks for the feedback as always!

fpringvaldsen commented:
@havocp I've addressed your feedback!

havocp commented:
Thanks!
